File: tests/WhiskerTracking/test_whiskertracker.cpp

#include "Geometry/lines.hpp"
#include "whiskertracker.hpp"

#include <catch2/catch_test_macros.hpp>
#include <catch2/benchmark/catch_benchmark.hpp>

#include "test_data.hpp"

TEST_CASE("Test Trace", "[Trace]") {

    static auto test_img = load_img("data/0110_1_0000000.bin");
    static auto wt = create_warmed_up_wt(test_img);

    wt.setWhiskerPad(577.0f, 202.0f);
    auto whiskers = wt.trace(test_img, img_height, img_width);

    CHECK(whiskers.size() == 5);

}

TEST_CASE("Test Trace Multi", "[Trace]") {

    static auto test_img = load_img("data/0110_1_0000000.bin");
    static auto test_img2 = load_img("data/test_img.bin");
    static auto wt = create_warmed_up_wt(test_img);

    wt.setWhiskerPad(577.0f, 202.0f);

    std::vector<std::vector<uint8_t>> images = {test_img};

    auto whiskers = wt.trace_multiple_images(images, img_height, img_width);

    CHECK(whiskers.size() == 1);
    CHECK(whiskers[0].size() == 5);

    std::vector<std::vector<uint8_t>> images2{
        load_img("data/0110_1_0000000.bin"),
        load_img("data/test_img.bin")};

    CHECK(images2.size() == 2);

    auto whiskers2 = wt.trace_multiple_images(images2, img_height, img_width);

    CHECK(whiskers2.size() == 2);
    CHECK(whiskers2[0].size() == 5);

    whiskers2 = wt.trace_multiple_images(images2, img_height, img_width);

    CHECK(whiskers2.size() == 2);
    CHECK(whiskers2[0].size() == 5);

}

File: tests/WhiskerTracking/test_whiskertracker.py

import pytest
import numpy as np
from whiskertracker import WhiskerTracker

def load_img(file_path):
    # Assuming the function loads an image from the file and returns it as a numpy array
    # This is a placeholder; replace with actual image loading logic
    return np.fromfile(file_path, dtype=np.uint8)

def create_warmed_up_wt(image):
    # Assuming this function initializes the WhiskerTracker with some pre-processing on the image
    # This is a placeholder; replace with actual initialization and pre-processing logic
    wt = WhiskerTracker()
    # Example pre-processing step
    wt.setWhiskerPad(577.0, 202.0)
    return wt

@pytest.fixture
def test_img(pytestconfig):
    dir_path = pytestconfig.getoption("dirpath")
    file_path = f'{dir_path}/0110_1_0000000.bin'
    return load_img(file_path)

@pytest.fixture
def wt(test_img):
    return create_warmed_up_wt(test_img)

def test_trace(wt, test_img):
    img_height, img_width = 480, 640  # Assuming these are the dimensions of the test image
    whiskers = wt.trace(test_img, img_height, img_width)
    assert len(whiskers) == 5